cpp/tests/test_abm.cpp

```diff
@@ -1421,111 +1445,98 @@ TEST(TestDiscreteDistribution, generate)
-                                                                    mio::abm::LocationType::Work};
-
-    auto testing_rule1 = mio::abm::TestingRule({}, test_location_types1, test_infection_states1);
-    std::vector<mio::abm::TestingRule> testing_rules = {testing_rule1};
-
-    const auto testing_frequency = mio::abm::days(1);
-    const auto start_date        = mio::abm::TimePoint(0);
-    const auto end_date          = mio::abm::TimePoint(60 * 60 * 24 * 3);
-    const auto probability       = 1.0;
-    const auto test_type         = mio::abm::PCRTest();
-
-    auto testing_scheme =
-        mio::abm::TestingScheme(testing_rules, testing_frequency, start_date, end_date, probability, test_type);
+    testing_scheme.update_activity_status(mio::abm::TimePoint(0));
 
     std::vector<mio::abm::InfectionState> test_infection_states2 = {mio::abm::InfectionState::Recovered_Carrier};
     std::vector<mio::abm::LocationType> test_location_types2     = {mio::abm::LocationType::Home};
-    auto testing_rule2 = mio::abm::TestingRule({}, test_location_types2, test_infection_states2);
-    testing_scheme.add_testing_rule(testing_rule2);
+    auto testing_criteria2 = mio::abm::TestingCriteria({}, test_location_types2, test_infection_states2);
+    testing_scheme.add_testing_criteria(testing_criteria2);
 
     auto loc_home = mio::abm::Location(mio::abm::LocationType::Home, 0);
-    auto loc_work = mio::abm::Location(mio::abm::LocationType::Home, 0);
+    auto loc_work = mio::abm::Location(mio::abm::LocationType::Work, 0);
     auto person1  = mio::abm::Person(loc_home, mio::abm::InfectionState::Carrier, mio::abm::AgeGroup::Age15to34, {});
     auto person2 =
         mio::abm::Person(loc_home, mio::abm::InfectionState::Recovered_Carrier, mio::abm::AgeGroup::Age15to34, {});
 
     ScopedMockDistribution<testing::StrictMock<MockDistribution<mio::UniformDistribution<double>>>> mock_uniform_dist;
     EXPECT_CALL(mock_uniform_dist.get_mock(), invoke)
-        .Times(testing::AtLeast(4))
+        .Times(testing::Exactly(5))
         .WillOnce(testing::Return(0.7))
         .WillOnce(testing::Return(0.5))
         .WillOnce(testing::Return(0.7))
-        .WillOnce(testing::Return(0.5));
-    ASSERT_EQ(testing_scheme.run_scheme(person1, loc_home), false);
-    ASSERT_EQ(testing_scheme.run_scheme(person2, loc_work), true);
-
-    ASSERT_EQ(person1.is_in_quarantine(), true);
-    ASSERT_EQ(person2.is_in_quarantine(), false);
-    ASSERT_EQ(person2.get_time_since_negative_test(), mio::abm::days(0));
+        .WillOnce(testing::Return(0.5))
+        .WillOnce(testing::Return(0.9));
+    ASSERT_EQ(testing_scheme.run_scheme(person1, loc_home), false); // Person tests and tests positive
+    ASSERT_EQ(testing_scheme.run_scheme(person2, loc_work), true); // Person tests and  tests negative
+    ASSERT_EQ(testing_scheme.run_scheme(person1, loc_home), true); // Person doesn't test
```

Comment:
I don't understand. Why does the person not test here? Because it tested positive shortly before? Then if return of run_scheme is if the person is allowed to enter or not, it should return the same as before, shouldnt it?

Reply:
Since the person just got tested and the minimal time of repeated testing is >0, it will not get tested again. Therefore, by the scheme, the person is now allowed to enter, since no test is performed.

Obviously, the person is in quarantine after the first test. Therefore, in a real scenario, a new location cannot be chosen by the MigrationRules. Checking for quarantine is NOT the task of the TestingScheme (at this point). This can be discussed.

Comment: To be allowed to enter a (new) location, ALL testing schemes have to allow the person to go there. If you would call run_strategy twice, and there is no testing scheme with minimal time = 0, then it could actually change the result from the first call. This has to be avoided. Right now, this is fine since run_strategy is only ever called once for each location (where quarantine has already be checked for in the MigrationRules).